doc/man/raduse.1
.\" $Id$
.ds RD /usr/local/etc/raddb
.ds RL /usr/adm
.ds RP /usr/local/etc/raddb
.TH raduse 1 "March 10, 2000" "FSF"
.SH NAME
raduse - monitor the usage of dialup lines.
.SH SYNOPSIS
\fBraduse\fR
[\fB-b\fR]
[\fB-d\fR count]
[\fB-D\fR]
[\fB-I\fR]
[\fB-i\fR|\fB-n\fR]
[\fB-s\fR delay]
[\fB-w\fR]
[NAS ...]
.PP
\fBraduse -l\fR
.PP
\fBraduse -h\fR
.SH DESCRIPTION
\fBRaduse\fR monitors the activity of dialup lines.
.PP
When used without arguments, it displays statistics about every dialup
line served by \fBRadius\fR server. Otherwise, each argument is
expected to be the name of the network access server for which the
information is to be displayed.
.PP
At the top of output the summary information is displayed. It consists
of two lines. First line shows the statistic collection uptime and
current date/time. Second line shows total number of lines, number of
active lines, number of idle (inactive) lines and load percentage.
.PP
The dialup statistics is displayed in the area below. For each dialup
line three lines of data are shown.
.PP
First line shows the network access server name, port number on that
server, number of logins registered on this line, status of the line,
amount of time the line keeps the current status, and date and time
where the line has switched to the current status.
.PP
If the line is currently active, the status field displays login
name of the user logged in on this line. If the line is inactive, the
word \fB[Idle]\fR is displayed.
.PP
Second and third lines display active and idle usage summary. They show
following data: total time the line is in the given state, maximum
amount of time in this state, and starting date and time when maximum
duration was perceived.

.SH OPTIONS
.IP -b
Start up in brief mode. In this mode only first line of information for
each dialup line is displayed.
.IP -d
Show only count displays, then exit. A display is considered to be one
update of the screen.
.IP -D
Dump the statistics database to the standard output and then
exit. This is for debugging purposes only.
.IP -I
Do not display idle lines. By default \fBraduse\fR displays all dialup
lines.
.IP -i
Use interactive mode. In this mode any input is immediately read for
processing. See the \fBCOMMANDS\fR section for the description of
commands usable in interactive mode. After processing each command the
screen is updated immediately, no matter was the command understood or
not. This mode is the default when the standard output is an
intelligent terminal.
.IP -n
Use non-interactive mode.
.IP -s
Specify delay in seconds between screen updates.
.IP -w
Widen the time display fields to show the seconds.
.IP -l
List the names and IP numbers of network access servers and then exit.
.IP -h
Display short usage summary.

.SH COMMANDS
When in interactive mode, \fBraduse\fR reads commands from the
terminal and acts upon them accordingly. Each command is either a
single letter or a combination control+letter. The command is
reacted upon immediately, unless it needs some additional information,
in which case the user will be prompted to enter such
information. While entering this information, erase and kill keys, as
set up by stty command, are recognized and a newline terminates the
input.
.PP
The following commands are understood (C-L means control-L):
.PP
.IP newline
Refresh the screen immediately
.IP space
Refresh the screen immediately
.IP C-L
Clear and redraw the display.
.IP ^
(Caret) go to the first page.
.IP b
Toggle brief display mode.
.IP C-b
Move one page backwards.
.IP C-f
Move one page forwards.
.IP i
Toggle idle line display on or off.
.IP j
Move one line forwards.
.IP k
Move one line backwards.
.IP G
Move to the last page.
.IP $
Move to the last page.
.IP q
Quit the program
.IP s
Change the number of seconds to delay between screen updates. Prompts
user to input the number desired.
.IP t
Display information about given Network Access Servers. Prompts user
to input the NAS names. The names should be separated by
whitespace. The word \fBall\fR returns to display of all NAS lines.

.SH FILES
.IP \*(RD/naslist
The list of network access servers

.IP \*(RL/radacct/radstat
Radius statistics.

.SH BUGS
The information in this manpage may be obsolete or incomplete. Please
refer to texinfo documentation for full information about GNU Radius
tools.

.SH AUTHORS
Sergey Poznyakoff, gray@farlep.net

.SH SEE ALSO
radiusd(8rad), radwho(1rad), radgrep(1rad), naslist(5rad)



NEWS
GNU Radius NEWS -- history of user-visible changes. 2004-04-23
Copyright (C) 2002,2003,2004 Free Software Foundation, Inc.
See the end of file for copying conditions.

Please send radius bug reports to <bug-gnu-radius@gnu.org>
^L
Version 1.2.3:

* libgnuradius

This is a library of functions for creation, handling and sending
requests using RADIUS protocol.

All programs have been rewritten to link with libgnuradius. On most
sites this will mean linking against a shared library, which will
reduce the size of the executables.

* gnuradius.scm

This is a guile module allowing to use libgnuradius functions. It
supercedes radscm program, which has been removed.

* Rewrite language

** Added i18n support

** New built-in functions:

*** Functions to access internal fields of a RADIUS request.
*** Interfaces to the Radius NAS database (raddb/naslist).
*** Interfaces to DNS lookup functions.

* radiusd

** Improved checking for multiple logins. Previous versions relied entirely
on the contents of /var/log/radutmp file. Starting at this version,
radiusd offers at least two methods of checking for multiple logins:
using the traditional radutmp file and using the SQL database. New
keywords has been added to the sqlserver file that declare the SQL
queries to be used when retrieving the information about currently
active sessions.

More methods of checking will be added in future versions.

** When an unsupported authentication type is requested, radiusd first
checks if an extension Scheme module is provided that handles that
authentication type. If such module is found, it is invoked to handle
the authentication.

** System accounting can be turned off by specifying `system no;' in
`acct' section of raddb/config. 

** New configuration statement 'load-module' allows to load arbitrary
Scheme modules.

** Detailed log filenames are configurable via `detail-file-name'
statements in `auth' and `acct' sections of raddb/config.

** Support for Guile versions prior to 1.6 has been withdrawn.

* Bugfixes

** Allow to omit port numbers in `listen' statements (raddb/config), as
described in the documentation.
    


Version 1.2:

* Important notes for upgrading from earlier versions of GNU Radius

** Upgrade your dictionaries.

This version does a very strict semantic checking on the dictionary
files. Old dictionaries may produce lots of warnings and
errors, therefore you are advised to upgrade them. To do so,
run `make update-dictionary'. It will update your main 
(raddb/dictionary) and secondary dictionaries (raddb/dict/*),
preserving any changes you did to the former. 

Notice, that, starting from this version, the default main dictionary
file will not contain any attribute declarations, instead it will
merely include the necessary dictionaries. However, you may add your
declarations to it and modify it to your linking.

Any modifications to the dictionary files in raddb/dict are now
strongly discouraged, as these files will be updated by each new
release. Put your modifications to the main dictionary file only.
This file will remain unchanged during subsequent installs.

Two new statements are provided to help you customize your dictionary.
See section `** Dictionary' below.

** Update your raddb/config and raddb/rewrite

The rewrite procedures are no longer stored in a single file
`radiusd/rewrite', instead they may occupy several source file.
The rewrite procedures shipped with this version of GNU Radius
are installed in ${prefix}/share/radius/1.2/rewrite. We strongly
advise you to add the following statement to your raddb/config:

rewrite {
        load "checknas.rw";
        load "log-hook.rw";
};

You may then remove the old raddb/rewrite file, *unless* you have
modified it.

If you have modified it, follow the procedure described below:

1. Load this file, by adding the following statement to the above
`rewrite' block:

	load "rewrite";

2. Make sure to remove from this file or rename the following functions:
check_snmp_u, check_snmp_s, decimal, check_max_finger, check_unix.

** Tailor log file format.

The format of log file is now fully customizable via a special Rewrite
procedures, called "hooks" (search for logging-hook in the text below
for the detailed description). To revert to logfile format used in
versions 1.0-1.1, add the following two statements to your raddb/config,
right after `logging {' line:

       	prefix-hook "compat_log_prefix";
	suffix-hook "compat_log_suffix";

** Using Guile extensions

If you are using Guile extensions, notice that the default location
for Radius scheme modules is now ${prefix}/share/radius/1.2/scheme.
	
* New features:

** Support for Status-Server request.

** Support for Tunnel Protocol (RFC 2865)

** The main configuration file raddb/config allows to specify several
Rewrite source files to be loaded instead of the single predefined
file raddb/rewrite.

** Arbitrary Rewrite expressions may be used in A/V pairs, for example:

DEFAULT	     Auth-Type = SQL
	     Framed-IP-Address =
	           "=10.10.10.1 + %[NAS-Port-Id]/16 + %[NAS-Port-Id]"

** Test shell mode considerably improved.

** The internal attributes do not appear in detailed logs. The
special flag 'l' has been added to the dictionary syntax to
turn this feature off. Thus, for example:

ATTRIBUTE Orig-User-Name  2101	string	- [L--R-R]l

will make Orig-User-Name attribute visible in detailed log files.

** Changes to Rewrite language

*** usual dotted-quad notation may be used to represent IP numbers.

*** a number of new built-in functions is added

*** #pragma regexp allows to alter the type of regular expressions
used. The rest of line following `#pragma regexp' must consist of
whitespace-separated regexp modifiers. Each regexp modifier is one
of the following keywords:

  extended      Use extended regular expressions
  icase         Ignore case
  newline       Match-any-character operators don't match a newline.

optionally preceded by plus, indicating enabling the feature, or

raduse/raduse.c
/* This file is part of GNU RADIUS.
   Copyright (C) 2000, Sergey Poznyakoff
  
   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.
  
   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.
  
   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software Foundation,
   Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA. */

#ifndef lint
static char rcsid[] = 
"$Id$";
#endif

#ifdef HAVE_CONFIG_H
# include <config.h>
#endif

#include <sys/types.h>
#include <sys/time.h>
#include <sys/file.h>
#include <sys/stat.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <errno.h>
#include <stdarg.h>
#include <ctype.h>
#include <netinet/in.h>
#if defined(sun)
# include <fcntl.h>
#endif
#include <sysdep.h>
#include <radius.h>
#include <radpaths.h>
#include <radutmp.h>
#include <radlast.h>
#include <log.h>

#include <getopt1.h>
#include "display.h"
#include "screen.h"

#define MAX_PORTS 256

#define C_B 2
#define C_F 6

typedef struct {
	int  use;
	int  nports;
	PORT_STAT *port;
} port_usage_t;

#define AP(p,m) (((port_usage_t*)(p)->app_data)-> ##m)

/* various options */
int width = 5;            /* width for time output (5 - hh:mm, 8 - hh:mm:ss) */
int delay = 1;            /* delay between screen updates (seconds) */
int maxscreens = -1;      /* max. number of screens to display */
int numscreens = 0;       /* number of screens displayed so far */
int brief = 0;            /* brief mode (don't show inuse/idle statistics) */
int show_idle = 1;        /* show idle lines */
int interactive = -1;     /* interactive mode */
int dump_only = 0;

char *file;
int nas_cnt;

time_t starttime;
unsigned offset;


void raduse();
int collect(int);
int update(int);
PORT_STAT * find_port(NAS *nas, int port_no);
void listnas();
void usage();
void license();
void add_login(NAS *nas, struct radutmp *bp);
void add_logout(NAS *nas, WTMP *pp, struct radutmp *bp);
void add_nas(char*);
void use_all();
void display();
void read_naslist();
void raduse();
void mark_all(int);
int mark_nas(char *);
void select_nas();
void dump(int);

#define OPTSTR "bd:DhIilLns:wx"

struct option longopt[] = {
	"brief",         no_argument,       0, 'b',
        "display",       required_argument, 0, 'd',
	"dump",          no_argument,       0, 'D',
	"no-idle-lines", no_argument,       0, 'I',
	"interactive",   no_argument,       0, 'i',
	"no-interactive",no_argument,       0, 'n',
	"delay",         required_argument, 0, 's',
	"widen",         no_argument,       0, 'w',
	"license",       no_argument,       0, 'L',
	"list-nas",      no_argument,       0, 'l',
	"help",          no_argument,       0, 'h',
	0,
};

/*
 * Debugging hook.
 * To debug, define symbol DEBUG, run raduse as
 *     raduse -x
 * and connect to it with gdb from another terminal.
 */
#if defined DEBUG
volatile int stop = 1;

void
wait_debug()
{
	printf("READY TO DEBUG: %d\n", getpid());
	while (!stop)
		;
}
#else
# define wait_debug()
#endif


int
main(argc, argv)
	int argc;
	char **argv;
{
	int c;

	app_setup();	
	initlog(argv[0]);
	while ((c = getopt_long(argc, argv, OPTSTR, longopt, NULL)) != EOF)
		switch(c) {
		case 'b':
			brief = !brief;
			break;
		case 'd':
			maxscreens = atoi(optarg);
			break;
		case 'D':
			dump_only++;
			break;
		case 'h':
			usage();
			/*NOTREACHED*/
		case 'I':
			show_idle = !show_idle;
			break;
		case 'i':
			interactive = 1;
			break;
		case 'L':
			license();
			return 0;
		case 'l':
			listnas();
			return 0;
		case 'n':
			interactive = 0;
			break;
		case 's':
			delay = atoi(optarg);
			break;
		case 'w':
			width = 8;
			break;
#ifdef DEBUG
		case 'x':
			stop = 0;
			break;
#endif			
		default:
			break;
		}

	radpath_init();
	read_naslist();
	file = radstat_path;

	if (optind == argc) {
		mark_all(1);
	} else {
		for (argv += optind; *argv; ++argv) 
			add_nas(*argv);
	}

	setlinebuf(stdout);
	
	setvbuf(stdout, NULL, _IONBF, 0);

	wait_debug();
	
	init_termcap(interactive);
	if (interactive == -1)
		interactive = smart_terminal;
	
	alloc_screen(nas_cnt, MAX_PORTS);
	init_screen();
	raduse();
	restore_screen();
	return 0;	
}

void
read_naslist()
{
	NAS *nas;
	char *file = mkfilename(radius_dir, RADIUS_NASLIST);
	if (nas_read_file(file) < 0)
		exit(1);
	efree(file);
	nas_cnt = 0;
	for (nas = nas_next(NULL); nas; nas = nas_next(nas)) {
		nas->app_data = emalloc(sizeof(port_usage_t));
		nas_cnt++;
	}
}

int
stat_insert_port(port)
	PORT_STAT *port;
{
	NAS *nas = nas_lookup_ip(port->ip);
	port_usage_t *pu;
	
	if (!nas) {
		radlog(L_ERR,
		    _("stat_insert_port(): portno %d: can't find nas for IP %s"),
		    port->port_no,
		    format_ipaddr(port->ip));
		free_entry(port);
		return -1;
	}

	pu = nas->app_data;
	if (pu->port == NULL) {
		pu->port = port;
	} else if (pu->port->port_no > port->port_no) {
		port->next = pu->port;
		pu->port = port;
	} else {
		PORT_STAT *p, *prevp = NULL;

		for (p = pu->port; p; prevp = p, p = p->next) {
			if (p->port_no > port->port_no)
				break;
		}
		port->next = p;
		prevp->next = port;
	}
	pu->nports++;
	return 0;
}


int
collect(fd)
	int fd;
{
	PORT_STAT stat;
	PORT_STAT *port;
	NAS *nas;
	
	if (lseek(fd, sizeof(PORT_STAT), SEEK_SET) != sizeof(PORT_STAT)) {
		radlog(L_ERR, _("lseek error on `%s' (%d): %s"),
		    file, sizeof(PORT_STAT), strerror(errno));
		return 1;
	}
	while (read(fd, &stat, sizeof(stat)) == sizeof(stat)) {
		if (stat.ip == 0)
			break;

		nas = nas_lookup_ip(stat.ip);
		if (!nas) {
			radlog(L_ERR,
			    _("collect(): port %d: can't find nas for IP %s"),
			    stat.port_no,
			    format_ipaddr(stat.ip));
			return 1;
		}

		port = find_port(nas, stat.port_no);
		if (!port) {
			port = alloc_entry(sizeof(*port));
			stat.next = 0;
			*port = stat;
			stat_insert_port(port);
		} else {
			port->active = stat.active;               
			strncmp(port->login, stat.login, sizeof(port->login));
			port->count = stat.count;                
			port->start = stat.start;
			port->lastin = stat.lastin;            
			port->lastout = stat.lastout;           
			port->inuse = stat.inuse;             
			port->idle = stat.idle;              
			port->maxinuse = stat.maxinuse;
			port->maxidle = stat.maxidle;
		}
	}
	return 0;
}

PORT_STAT *
find_port(nas, port_no)
	NAS *nas;
	int port_no;
{
	PORT_STAT *port;
	
	/* First try to find it in the cached buffer */
	for (port = AP(nas,port); port; port = port->next) 
		if (port->port_no == port_no)
			return port;
	return NULL;
}

void
raduse()
{
	int wfd;
	struct stat stb;
	PORT_STAT stat;
	struct timeval timeout;
	fd_set rfd;
	char cmd;
	time_t lastmtime;
	
	wfd = open(file, O_RDONLY);
	if (wfd == -1) {
		radlog(L_ERR|L_PERROR, _("can't open `%s'"), file);
		return;
	}
	
	/* Read header record */
	if (read(wfd, &stat, sizeof(stat)) != sizeof(stat)) {
		radlog(L_ERR, _("read error on `%s': %s"),
		    file, strerror(errno));
		return;
	}
	starttime = stat.start;
	fstat(wfd, &stb);
	lastmtime = stb.st_mtime;
	offset = sizeof(stat);
	
	if (dump_only) {
		dump(wfd);
		return;
	}
	
	if (collect(wfd))
		return;
	
	while (maxscreens == -1 || (numscreens++ < maxscreens)) {
		display();

		FD_ZERO(&rfd);
		FD_SET(1, &rfd);
		timeout.tv_sec = delay;
		timeout.tv_usec = 0;

/*		fstat(wfd, &stb);
		if (stb.st_mtime > lastmtime) {
			collect(wfd);
		}
*/
		collect(wfd);
		
		if (interactive &&
		    select(32, &rfd, NULL, NULL, &timeout) > 0) {
			read(0, &cmd, 1);
			switch (cmd) {
			case '\n':
			case ' ':
				/* refresh */
				clearmsg();
				break;
			case '\f':
				clear();
				display();
				break;
			case '^':
				firstpage();
				break;
			case 'b':
				brief = !brief;
				if (brief)
					clear();
				break;
			case C_B:
				page(-1);
				break;
			case C_F:
				page(1);
				break;
			case 'i':
				show_idle = !show_idle;
				if (!show_idle)
					clear();
				break;
			case 'j': /* Next record */
				/*scroll(brief ? 1 : 3);*/
				scroll(1);
				break;
			case 'G':
			case '$':
				lastpage();
				break;
			case 'k': /* Prev record */
				/*scroll(brief ? -1 : -3);*/
				scroll(-1);
				break;
			case 'q': /* ie. stop */
				numscreens = maxscreens = 0; 
				break;
			case 's':
				getint(_("Seconds to delay: "), &delay);
				break;
			case 't':
				select_nas();
				break;
			default:
				msg(MT_standout,
				    _("unknown command: %c"), cmd);
			}
		}
	}
	close(wfd);
}

int
formatdelta(outbuf, delta)
	char *outbuf;
	time_t delta;
{
	char ct[128];
	char buf[128];
	struct tm *tm;

	tm = gmtime(&delta);
	strftime(ct, sizeof(ct), "%c", tm);
	if (delta < 86400)
		sprintf(buf, "%*.*s", width, width, ct + 11);
	else
		sprintf(buf, "%ld+%*.*s",
			delta / 86400, width, width, ct + 11);
	return sprintf(outbuf, "%11.11s ", buf);
}

int
formattime(buf, time)
	char *buf;
	time_t time;
{
	struct tm *tm;
	char ct[128];
	
	tm = localtime(&time);
	strftime(ct, sizeof(ct), "%c", tm);
	return sprintf(buf, "%10.10s %5.5s ", ct, ct + 11);

	/*"%m/%d/%y %H:%M:%S", tm);
	printf("%8.8s %5.5s ", buf, buf + 9);*/
}

int
formatstop(buf, time)
	char *buf;
	time_t time;
{
	struct tm *tm;
	char ct[128];
	
	tm = localtime(&time);
	strftime(ct, sizeof(ct), "%H:%M:%S", tm);
	return sprintf(buf, "%5.5s ", ct);
}
	
void
display()
{
	NAS *nas;
	PORT_STAT *port;
	int j, off = 0;
	time_t now = time(NULL), delta, stop;
	char *str;
	int total_lines = 0, active_lines = 0;
	
	for (nas = nas_next(NULL); nas; nas = nas_next(nas)) {
		for (port = AP(nas,port); port; port = port->next) {
			total_lines++;
			if (port->active)
				active_lines++;
		}
	}

	
	str = headerbuf[0];
	str += sprintf(str, _("uptime "));
	str += formatdelta(str, now - starttime);
	str += sprintf(str, "        ");
	formattime(str, now);
	
	str = headerbuf[1];
	str += sprintf(str, _("%3d lines, %3d active, %3d idle. "),
		       total_lines,
		       active_lines,
		       total_lines - active_lines);
	if (total_lines)
		str += sprintf(str, _("Pool load %4.2f"),
			       (double) active_lines / total_lines);
	else
		str += sprintf(str, _("Pool load ??.??"));

	off = 0;
	for (nas = nas_next(NULL); nas; nas = nas_next(nas)) {
		if (!AP(nas,use))
			continue;
		j = 0;
		for (port = AP(nas,port); port; port = port->next) {
			if (!show_idle && !port->active) 
				continue;
			/* Port number */
			str = screen[off + j++];
			str += sprintf(str, "%-12.12s %3.3d %5d ",
				      nas->shortname,
				      port->port_no, port->count);
			if (port->active) {
				delta = now - port->lastin;
				str += sprintf(str,
					       "%-16.16s ", port->login);
				str += formatdelta(str, delta);
				str += formattime(str, port->lastin);
			} else {
				/* currently idle */
				delta = now - port->lastout;
				str += sprintf(str,
					       "%-16.16s ", "[idle]");
				str += formatdelta(str, delta);
				str += formattime(str, port->lastout);
			}

			if (brief)
				continue;
			str = screen[off + j++];
			str += sprintf(str, "          ");
			/* In use statistics */
			str += formatdelta(str, port->inuse);
			str += formatdelta(str, port->maxinuse.time);
			str += formattime(str, port->maxinuse.start);
			str += sprintf(str, "- ");
			stop = port->maxinuse.start +
				port->maxinuse.time;
			str += formatstop(str, stop);

			/* Port number */
			str = screen[off + j++];
			str += sprintf(str, "          ");

			/* Idle time statistics */
			str += formatdelta(str, port->idle);
			str += formatdelta(str, port->maxidle.time);
			str += formattime(str, port->maxidle.start);
			str += sprintf(str, "- ");
			stop = port->maxidle.start +
				port->maxidle.time;
			str += formatstop(str, stop);
		}
		off += j;
	}
	update_display(off);
}

void
add_nas(name)
	char *name;
{
	NAS *nas;
	
	nas = nas_lookup_name(name);
	if (!nas) {
		radlog(L_ERR, _("no such NAS: %s (use raduse -l to get the list)"),
			name);
		exit(1);
	} else
		AP(nas,use) = 1;
}

void
listnas()
{
	NAS *p;
	
	for (p = nas_next(NULL); p; p = nas_next(p)) {
		printf("%-32.32s %-10.10s %-16.16s\n",
		       p->longname,
		       p->shortname,
		       format_ipaddr(p->ipaddr));
	}
}

void
mark_all(m)
	int m;
{
	NAS *nas;

	for (nas = nas_next(NULL); nas; nas = nas_next(nas))
		AP(nas,use) = m;
}

int
mark_nas(name)
	char *name;
{
	NAS *nas;

	for (nas = nas_next(NULL); nas; nas = nas_next(nas))
		if (strcasecmp(nas->shortname, name) == 0) {
			AP(nas,use) = 1;
			return 0;
		}
	return 1;
}
	
void
select_nas()
{
	char buf[80];
	int nas_cnt = 0;
	char *p, *nasname;

#define PROMPT _("NASes to show:")       
	msg(MT_standout, PROMPT);

	if (readline(buf, sizeof(buf)-sizeof(PROMPT), 0) > 0) {
		mark_all(0);
		p = buf;
		do {
			while (*p && isspace(*p))
				p++;
			if (!*p)
				break;
			nasname = p;
 
			while (*p && !isspace(*p))
				p++;
			if (*p)
				*p++ = 0;
			nas_cnt++;
			if (strcasecmp(nasname, "all") == 0) {
				mark_all(1);
				break;
			} else if (mark_nas(nasname)) {
				msg(MT_standout, _("No such NAS: ``%s'"), nasname);
				mark_all(1);
				return;
			}
		} while (*p);
	}
	if (!nas_cnt) 
		mark_all(1);
	clearmsg();
}

void
dump(fd)
	int fd;
{
	PORT_STAT stat;
	NAS *nas;
	unsigned off = 0;
	
	while (read(fd, &stat, sizeof(stat)) == sizeof(stat)) {
		if (stat.ip != 0) {
			nas = nas_lookup_ip(stat.ip);
			printf("%8d %16.16s %4d\n",
			       off,
			       nas->shortname,
			       stat.port_no);
		}
		off += sizeof(stat);
	}
	return;
}




char usage_str[] = 
"usage: raduse [options] [nas [nas...]]\n"
"The options are:\n"
"       -b, --brief               Brief mode.\n"
"       -d, --display COUNT       Show only COUNT displays.\n"
"	-D, --dump                Dump the statistics database to stdout\n"
"                                 and exit.\n"
"	-I, --no-idle-lines       Do not display idle lines.\n"
"	-i, --interactive         Interactive mode.\n"
"	-n, --no-interactive      Non-interactive mode.\n"
"	-s, --delay NUM           Specify delay in seconds between\n"
"                                 screen updates.\n"
"	-w, --widen               Widen the time display fields to show\n"
"                                 the seconds.\n"
"	-l, --list-nas            List the names and IP numbers of\n"
"                                 network access servers and then exit.\n"
"       -L, --license             Display GNU lisense and exit\n"
"	-h, --help                Display short usage summary.\n";

void 
usage()
{
	printf("%s", usage_str);
	printf("\nReport bugs to <%s>\n", bug_report_address);
	exit(1);
}

static char license_text[] = "\
\n\
  This program is free software; you can redistribute it and/or modify\n\
  it under the terms of the GNU General Public License as published by\n\
  the Free Software Foundation; either version 2 of the License, or\n\
  (at your option) any later version.\n\
\n\
  This program is distributed in the hope that it will be useful,\n\
  but WITHOUT ANY WARRANTY; without even the implied warranty of\n\
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the\n\
  GNU General Public License for more details.\n\
 \n\
  You should have received a copy of the GNU General Public License\n\
  along with this program; if not, write to the Free Software\n\
  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.\n\
";

void
license()
{
	printf("%s: Copyright 1999,2000 Sergey Poznyakoff\n", progname);
	printf("\nThis program is part of GNU Radius\n");
	printf("%s", license_text);
	exit(0);
}
